### DISCUSSION TITLE
Whitelist a request

### DISCUSSION THREAD

**Discussion by rosshulford** [mswjs/msw#1231] [Q&A] [answered]
I have a couple of requests I need to be ignored by MSW - is there a way to achieve this?

**Comment by kettanaito** [comment, marked as answer]
Hey, @rosshulford.

## Passthrough by default

MSW only captures those requests which you describe in your [request handlers](https://mswjs.io/docs/basics/request-handler). All other requests won't be affected and will be performed as-is. 

In the browser, all requests will go through the Service Worker regardless if they should be mocked or not. This is an implementation detail of the library and should not affect you in any way (the original responses will be returned in case a request doesn't have a corresponding request handler on your side). 

## Whitelist for `onUnhandledRequest`

By default, MSW will try to warn you when a passthrough request happens. This is an optimal DX in regards to people forgetting to mock requests or making typos in their handlers. 

Sometimes, there are cases when you would want MSW to stop warning you on certain unhandled requests (i.e. static assets requests or webpack HMR requests). To do that, provide a custom callback to the `onUnhandledRequest` function:

```js
worker.start({
  onUnhandledRequest(req, print) {
    if (req.url.pathname.startsWith('/assets/') {
      return 
    }

    print.warning()
  }
})
```

> The usage above will prevent MSW from warning you about unhandled requests that start with `/assets/`. 

Note that `onUnhandledRequest` has no effect over mocks whatsoever.

**Reply by kettanaito** [reply]
I'm marking this as the answer because it goes at length about whitelisting with MSW. You can ask follow-up questions if this answer isn't enough.

**Reply by cmacdonnacha** [reply]
Great answer but I have one follow up question. What types should we give `req` and `print`?

Also, is there a way to ignore multiple things like `'/assets/'` AND `googlefonts` etc. So basically an array of things to ignore.

**Reply by kettanaito** [reply]
> What types should we give req and print?

Their types are inferred. If you're abstracting them in your own function, here are the types: 

https://github.com/mswjs/msw/blob/1b9dde0193bf170c9ed0b89dee6226021e067f76/src/utils/request/onUnhandledRequest.ts#L18-L26

I guess we are not exporting them, right? Do you mind opening a pull request if that's the case?

**Reply by cmacdonnacha** [reply]
Yea they're not being inferred currently.

![image](https://user-images.githubusercontent.com/9777763/171268227-c5f816da-7533-46f1-93c5-458ecb6d4ebd.png)

**Reply by kettanaito** [reply]
Okay that sounds wrong. Can you please send me a reproduction I can see this?

**Reply by nicholaspretorius** [reply]
@cmacdonnacha To your second question: 

> Also, is there a way to ignore multiple things like '/assets/' AND googlefonts etc. So basically an array of things to ignore.

Using [array.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) worked for me: 

```
worker.start({
    // turn off MSW warnings for specific routes 
    onUnhandledRequest(req, print) {
      // specify routes to exclude
      const excludedRoutes = ['/favicon.ico', '/manifest.json', '/logo192.png'];

      // check if the req.url.pathname contains excludedRoutes
      const isExcluded = excludedRoutes.some(route => req.url.pathname.includes(route));
    
      if (isExcluded) {
        return; 
      }
  
      print.warning()
    }
  });
```

Screenshot of browser console _before_ exclusions: 

![msw-before-exclusions](https://user-images.githubusercontent.com/72647687/215260185-8259a337-ec61-408b-ab8d-0fa5e83bee7c.png)

Screenshot of browser console _after_ exclusions: 

![msw-after-exclusions](https://user-images.githubusercontent.com/72647687/215260195-983401f7-2716-42a4-96bf-02eb7912ff7f.png)

**Comment by rosshulford** [comment]
Many thanks.

**Reply by Dmo16** [reply]
Should this work with secured endpoints?

For some reason I have several `https` endpoints that I can't force an ignore on.

An example of a few of them:

`https://model-bird-11.clerk.accounts.dev/v1/client`

`https://model-bird-11.clerk.accounts.dev/v1/someotherendpoint`

My setup file:

```
const ignoredPathnames = [
  '/model-bird-11.clerk.accounts.dev',
  '/kit.fontawesome.com',
  '/_next/static/webpack',
  '/favicon.ico',
  '/_next/static/media',
  '/images.clerk.dev',
];

async function initMocks() {
  if (typeof window === 'undefined') {
    const { server } = await import('./server');
    server.listen();
  } else {
    const { worker } = await import('./browser');
    await worker.start({
      onUnhandledRequest(req, print) {
        if (ignoredPathnames.some((pathname) => req.url.pathname.startsWith(pathname))) {
          return;
        }
        print.warning();
      },
    });
  }
}

initMocks().then((r) => r);

export {};
```

Everything else in that array successfully gets ignored.

**Reply by kettanaito** [reply]
@Dmo16, the issue must be somewhere in your `if` condition. MSW doesn't treat HTTPS requests any differently from HTTP—both will arrive at the `onUnhandledRequest` (given they, indeed, don't have any corresponding request handlers so watch out for those too).

**Comment by timsofteng** [comment]
@kettanaito may I hide unhandled requests from browser's devtools as well? It is quite annoying too have a lot of unnecessary requests such as chunks and assets in network tab.

**Reply by kettanaito** [reply]
I'm not sure what you intend here. An unhandled request is the one that has no request handlers in MSW. If that's the case, _it will be performed as-is_, which means it will produce a network entry. 

A request won't appear in the Network only if it never happened. 

Do you perchance mean to hide the MSW's console messages about the unhandled requests? If that's the case, you can control the strategy used to notify you about them using the [`onUnhandledRequest`](https://mswjs.io/docs/api/setup-worker/start#onunhandledrequest) option on the `worker`. I showcased how to use it to ignore certain unhandled requests while warn/error on others in https://github.com/mswjs/msw/discussions/1231#discussioncomment-2729999.

**Reply by timsofteng** [reply]
>  it will be performed as-is, which means it will produce a network entry.

As I can see it duplicate unhandled requests in devtool's network tab.

![image](https://github.com/mswjs/msw/assets/51912173/13a5dc11-4fb2-4349-8f1c-c93910656b35)


In addition I see some unexpected requests. Take a lookt at it please

![image](https://github.com/mswjs/msw/assets/51912173/18a38897-fe01-48be-8834-e04ab4e1d0c5)
I don't see such requests in devtools without msw.

In addition msw shows some chunks imports which I would like to ignore in devtools according to xhr filter
<img width="641" alt="image" src="https://github.com/mswjs/msw/assets/51912173/cbae2f66-598d-4c07-87b3-3494f291cce8">

**Reply by timsofteng** [reply]
@kettanaito At the result for initator === mockServiceWorker I would like to see only handeled requests which I've described in `setupWorker(handlers)` and nothing more.

**Reply by kettanaito** [reply]
You see those requests not because they are unhandled but because MSW is enabled. When using MSW, all requests go through the Service Worker, which results in you seeing "duplicate" requests:

- `/resource`
- ⚙️ `/resource` (from the Service Worker)

**Those are not duplicates**. The request is still sent _once_ and processed once. What happens is this:

- `/resource` is dispatched in your app. You see it as the plain `/resource` network entry.
- The Service Worker intercepted that request, which creates a ⚙️ `/resource` network entry. This doesn't indicate a new request, just that it has been handled by the worker.
- From here, the worker can either respond with a mock or bypass the request. Either case will resolve the entry with the cogwheel and send the response to the original `/resource` network entry.

There is no way to opt-out of this by design. Not having those worker network entries would mean not sending those requests to the worker. To decide that, MSW needs to do request matching. Request matching requires an intercepted request instance. If you want to do that _before_ the request reaches the worker, it means procuring that request interception by other means, which contradicts the entire purpose of this library. 

TL;DR ignore this, this is a Service Worker's behavior.

**Reply by jamiehaywood** [reply]
I wanted to add something here in case it was of use to anyone irritated by the number of duplicated network entries (⚙️) in the chrome network tab.
Add `-is:service-worker-initiated` as a filter to your network tab, it will only show the XHR requests, which is supremely useful for less cluttered network debugging.

**Reply by isimmons** [reply]
path for external urls like "https://graphql-pokemon2.vercel.app" is "/" so it will return false. You can add the additional check for url.host in the excludedRoutes function. Also creating a new URL from req.url makes Typescript happy.

In this example 'img/pokemon'  is local (public dir) and matches url.pathname while 'graphql-pokemon2.vercel.app' is external and matches url.host

```javascript
    onUnhandledRequest(req, print) {
      const url = new URL(req.url)
      const excludedRoutes = ['img/pokemon', 'graphql-pokemon2.vercel.app']

      const isExcluded = excludedRoutes.some((route) => {
        return url.pathname.includes(route) || url.host.includes(route)
      })
      
      if (isExcluded) {
        return
      }

      print.warning()
    }
```
